File: sdks/python/hatchet_sdk/features/workflows.py
import asyncio
from datetime import timedelta
from typing import Literal

from hatchet_sdk.clients.rest.api.workflow_api import WorkflowApi
from hatchet_sdk.clients.rest.api.workflow_run_api import WorkflowRunApi
from hatchet_sdk.clients.rest.api_client import ApiClient
from hatchet_sdk.clients.rest.models.pause_workflow_request import PauseWorkflowRequest
from hatchet_sdk.clients.rest.models.pause_workflow_request_pause import (
    PauseWorkflowRequestPause,
)
from hatchet_sdk.clients.rest.models.pause_workflow_request_unpause import (
    PauseWorkflowRequestUnpause,
)
from hatchet_sdk.clients.rest.models.workflow import Workflow
from hatchet_sdk.clients.rest.models.workflow_list import WorkflowList
from hatchet_sdk.clients.rest.models.workflow_pause_scheduled_cron_run_queue_behavior import (
    WorkflowPauseScheduledCronRunQueueBehavior,
)
from hatchet_sdk.clients.rest.models.workflow_update_request import (
    WorkflowUpdateRequest,
)
from hatchet_sdk.clients.rest.models.workflow_version import WorkflowVersion
from hatchet_sdk.clients.rest.tenacity_utils import tenacity_retry
from hatchet_sdk.clients.v1.api_client import BaseRestClient
from hatchet_sdk.utils.timedelta_to_expression import (
    timedelta_to_expr,
)


class WorkflowsClient(BaseRestClient):
    """
    The workflows client is a client for managing workflows programmatically within Hatchet.

    Note that workflows are the declaration, _not_ the individual runs. If you're looking for runs, use the `RunsClient` instead.
    """

    def _wra(self, client: ApiClient) -> WorkflowRunApi:
        return WorkflowRunApi(client)

    def _wa(self, client: ApiClient) -> WorkflowApi:
        return WorkflowApi(client)

    async def aio_get(self, workflow_id: str) -> Workflow:
        """
        Get a workflow by its ID.

        :param workflow_id: The ID of the workflow to retrieve.
        :return: The workflow.
        """
        return await asyncio.to_thread(self.get, workflow_id)

    def get(self, workflow_id: str) -> Workflow:
        """
        Get a workflow by its ID.

        :param workflow_id: The ID of the workflow to retrieve.
        :return: The workflow.
        """
        with self.client() as client:
            workflow_get = tenacity_retry(
                self._wa(client).workflow_get, self.client_config.tenacity
            )
            return workflow_get(workflow_id)

    def list(
        self,
        workflow_name: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> WorkflowList:
        """
        List all workflows in the tenant determined by the client config that match optional filters.

        :param workflow_name: The name of the workflow to filter by.
        :param limit: The maximum number of items to return.
        :param offset: The offset to start the list from.

        :return: A list of workflows.
        """
        with self.client() as client:
            workflow_list = tenacity_retry(
                self._wa(client).workflow_list, self.client_config.tenacity
            )
            return workflow_list(
                tenant=self.client_config.tenant_id,
                limit=limit,
                offset=offset,
                name=self.client_config.apply_namespace(workflow_name),
            )

    async def aio_list(
        self,
        workflow_name: str | None = None,
        limit: int | None = None,
        offset: int | None = None,
    ) -> WorkflowList:
        """
        List all workflows in the tenant determined by the client config that match optional filters.

        :param workflow_name: The name of the workflow to filter by.
        :param limit: The maximum number of items to return.
        :param offset: The offset to start the list from.

        :return: A list of workflows.
        """
        return await asyncio.to_thread(self.list, workflow_name, limit, offset)

    def get_version(
        self, workflow_id: str, version: str | None = None
    ) -> WorkflowVersion:
        """
        Get a workflow version by the workflow ID and an optional version.

        :param workflow_id: The ID of the workflow to retrieve the version for.
        :param version: The version of the workflow to retrieve. If None, the latest version is returned.
        :return: The workflow version.
        """
        with self.client() as client:
            workflow_get_version = tenacity_retry(
                self._wa(client).workflow_version_get, self.client_config.tenacity
            )
            return workflow_get_version(workflow_id, version)

    async def aio_get_version(
        self, workflow_id: str, version: str | None = None
    ) -> WorkflowVersion:
        """
        Get a workflow version by the workflow ID and an optional version.

        :param workflow_id: The ID of the workflow to retrieve the version for.
        :param version: The version of the workflow to retrieve. If None, the latest version is returned.
        :return: The workflow version.
        """
        return await asyncio.to_thread(self.get_version, workflow_id, version)

    def delete(self, workflow_id: str) -> None:
        """
        Permanently delete a workflow.

        **DANGEROUS: This will delete a workflow and all of its data**

        :param workflow_id: The ID of the workflow to delete.
        :return: None
        """

        with self.client() as client:
            return self._wa(client).workflow_delete(workflow_id)

    async def aio_delete(self, workflow_id: str) -> None:
        """
        Permanently delete a workflow.

        **DANGEROUS: This will delete a workflow and all of its data**

        :param workflow_id: The ID of the workflow to delete.
        :return: None
        """

        return await asyncio.to_thread(self.delete, workflow_id)

    def pause(
        self,
        workflow_id: str,
        queue_ttl: timedelta,
        paused_workflow_cron_run_queue_behavior: Literal["DROP", "QUEUE"] = "QUEUE",
        paused_workflow_scheduled_run_queue_behavior: Literal[
            "DROP", "QUEUE"
        ] = "QUEUE",
    ) -> Workflow:
        """
        Pause a workflow.

        :param workflow_id: The ID of the workflow to pause.
        :param queue_ttl: The TTL for queued runs while the workflow is paused before they get dropped.
        :param paused_workflow_cron_run_queue_behavior: The behavior of cron runs triggered while the workflow is paused.
        :param paused_workflow_scheduled_run_queue_behavior: The behavior of scheduled runs triggered while the workflow is paused.
        :return: The updated workflow.
        """
        ttl_expr = timedelta_to_expr(queue_ttl)

        with self.client() as client:
            workflow_update_pause = tenacity_retry(
                self._wa(client).workflow_update, self.client_config.tenacity
            )
            return workflow_update_pause(
                workflow_id,
                WorkflowUpdateRequest(
                    pause=PauseWorkflowRequest(
                        PauseWorkflowRequestPause(
                            action="pause",
                            pausedWorkflowCronRunQueueBehavior=WorkflowPauseScheduledCronRunQueueBehavior(
                                paused_workflow_cron_run_queue_behavior
                            ),
                            pausedWorkflowScheduledRunQueueBehavior=WorkflowPauseScheduledCronRunQueueBehavior(
                                paused_workflow_scheduled_run_queue_behavior
                            ),
                            pausedWorkflowQueueTTL=ttl_expr,
                        )
                    )
                ),
            )

    async def aio_pause(
        self,
        workflow_id: str,
        queue_ttl: timedelta,
        paused_workflow_cron_run_queue_behavior: Literal["DROP", "QUEUE"] = "QUEUE",
        paused_workflow_scheduled_run_queue_behavior: Literal[
            "DROP", "QUEUE"
        ] = "QUEUE",
    ) -> Workflow:
        """
        Pause a workflow.

        :param workflow_id: The ID of the workflow to pause.
        :param queue_ttl: The TTL for queued runs while the workflow is paused before they get dropped.
        :param paused_workflow_cron_run_queue_behavior: The behavior of cron runs triggered while the workflow is paused.
        :param paused_workflow_scheduled_run_queue_behavior: The behavior of scheduled runs triggered while the workflow is paused.
        :return: The updated workflow.
        """

        return await asyncio.to_thread(
            self.pause,
            workflow_id,
            queue_ttl,
            paused_workflow_cron_run_queue_behavior,
            paused_workflow_scheduled_run_queue_behavior,
        )

    def unpause(self, workflow_id: str) -> Workflow:
        """
        Unpause a workflow.

        :param workflow_id: The ID of the workflow to unpause.
        :return: The updated workflow.
        """
        with self.client() as client:
            workflow_update_pause = tenacity_retry(
                self._wa(client).workflow_update, self.client_config.tenacity
            )
            return workflow_update_pause(
                workflow_id,
                WorkflowUpdateRequest(
                    pause=PauseWorkflowRequest(
                        PauseWorkflowRequestUnpause(action="unpause")
                    )
                ),
            )

    async def aio_unpause(self, workflow_id: str) -> Workflow:
        """
        Unpause a workflow.

        :param workflow_id: The ID of the workflow to unpause.
        :return: The updated workflow.
        """

        return await asyncio.to_thread(self.unpause, workflow_id)

File: sdks/ruby/src/lib/hatchet/features/workflows.rb
# frozen_string_literal: true

module Hatchet
  module Features
    # Workflows client for managing workflow definitions within Hatchet
    #
    # Note that workflows are the declaration, _not_ the individual runs.
    # If you're looking for runs, use the Runs client instead.
    #
    # @example Getting a workflow
    #   workflow = hatchet.workflows.get("workflow-id")
    #
    # @example Listing workflows
    #   workflows = hatchet.workflows.list(workflow_name: "my-workflow", limit: 10)
    #
    # @since 0.1.0
    class Workflows
      # Initializes a new Workflows client instance
      #
      # @param rest_client [Object] The configured REST client for API communication
      # @param config [Hatchet::Config] The Hatchet configuration containing tenant_id and other settings
      # @return [void]
      # @since 0.1.0
      def initialize(rest_client, config)
        @rest_client = rest_client
        @config = config
        @workflow_api = HatchetSdkRest::WorkflowApi.new(rest_client)
      end

      # Get a workflow by its ID
      #
      # @param workflow_id [String] The ID of the workflow to retrieve
      # @return [Object] The workflow details
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   workflow = hatchet.workflows.get("workflow-123")
      def get(workflow_id)
        @workflow_api.workflow_get(workflow_id)
      end

      # List all workflows in the tenant matching optional filters
      #
      # @param workflow_name [String, nil] The name of the workflow to filter by (namespace will be applied)
      # @param limit [Integer, nil] The maximum number of items to return
      # @param offset [Integer, nil] The offset to start the list from
      # @return [Object] A list of workflows
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   workflows = hatchet.workflows.list(workflow_name: "my-workflow", limit: 10, offset: 0)
      def list(workflow_name: nil, limit: nil, offset: nil)
        @workflow_api.workflow_list(
          @config.tenant_id,
          {
            limit: limit,
            offset: offset,
            name: workflow_name ? @config.apply_namespace(workflow_name) : nil,
          },
        )
      end

      # Get a workflow version by the workflow ID and an optional version
      #
      # @param workflow_id [String] The ID of the workflow to retrieve the version for
      # @param version [String, nil] The version to retrieve. If nil, the latest version is returned
      # @return [Object] The workflow version
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   version = hatchet.workflows.get_version("workflow-123", version: "v2")
      def get_version(workflow_id, version: nil)
        @workflow_api.workflow_version_get(workflow_id, { version: version })
      end

      # Permanently delete a workflow
      #
      # **DANGEROUS: This will delete a workflow and all of its data**
      #
      # @param workflow_id [String] The ID of the workflow to delete
      # @return [void]
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   hatchet.workflows.delete("workflow-123")
      def delete(workflow_id)
        @workflow_api.workflow_delete(workflow_id)
      end

      # Pause a workflow. While paused, new runs of the workflow are queued but not started.
      #
      # @param workflow_id [String] The ID of the workflow to pause
      # @param queue_ttl [Integer, String] How long runs stay queued while the workflow is paused before they are dropped, in seconds or as a duration string (e.g. "1h30m")
      # @param paused_workflow_cron_run_queue_behavior [String] The behavior of cron runs triggered while the workflow is paused ("QUEUE" or "DROP")
      # @param paused_workflow_scheduled_run_queue_behavior [String] The behavior of scheduled runs triggered while the workflow is paused ("QUEUE" or "DROP")
      # @return [HatchetSdkRest::Workflow] The updated workflow
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   hatchet.workflows.pause("workflow-123", queue_ttl: 3600)
      def pause(
        workflow_id,
        queue_ttl:,
        paused_workflow_cron_run_queue_behavior: HatchetSdkRest::WorkflowPauseScheduledCronRunQueueBehavior::QUEUE,
        paused_workflow_scheduled_run_queue_behavior: HatchetSdkRest::WorkflowPauseScheduledCronRunQueueBehavior::QUEUE
      )
        pause_request = HatchetSdkRest::PauseWorkflowRequestPause.new(
          action: "pause",
          paused_workflow_queue_ttl: queue_ttl.is_a?(String) ? queue_ttl : "#{queue_ttl}s",
          paused_workflow_cron_run_queue_behavior: paused_workflow_cron_run_queue_behavior,
          paused_workflow_scheduled_run_queue_behavior: paused_workflow_scheduled_run_queue_behavior,
        )

        @workflow_api.workflow_update(
          workflow_id,
          HatchetSdkRest::WorkflowUpdateRequest.new(pause: pause_request),
        )
      end

      # Unpause a workflow
      #
      # @param workflow_id [String] The ID of the workflow to unpause
      # @return [HatchetSdkRest::Workflow] The updated workflow
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   hatchet.workflows.unpause("workflow-123")
      def unpause(workflow_id)
        @workflow_api.workflow_update(
          workflow_id,
          HatchetSdkRest::WorkflowUpdateRequest.new(
            pause: HatchetSdkRest::PauseWorkflowRequestUnpause.new(action: "unpause"),
          ),
        )
      end
    end
  end
end
